amaranth/vendor/gowin.py

```diff
@@ -0,0 +1,549 @@
+                add_file {{file}}
+            {% endfor %}
+            add_file -type verilog {{name}}.v
+            add_file -type cst {{name}}.cst
+            add_file -type sdc {{name}}.sdc
+            set_device -name {{platform._device_name}} {{platform._part}}
+            set_option -verilog_std v2001 -print_all_synthesis_warning 1 -show_all_warn 1
+            {{get_override("add_options")|default("# (add_options placeholder)")}}
+            run all
+        """,
+    }
+
+    _gowin_command_templates = [
+        r"""
+        {{invoke_tool("gw_sh")}}
+            {{name}}.tcl
+        """,
+        "cp impl/pnr/project.fs {{name}}.fs"
+    ]
+
+    def __init__(self, *, toolchain="Apicula"):
+        super().__init__()
+
+        assert toolchain in ("Apicula", "Gowin")
+        self.toolchain = toolchain
+
+    @property
+    def _toolchain_env_var(self):
+        if self.toolchain == "Apicula" or self.toolchain == "Gowin":
+            return f"AMARANTH_ENV_{self.toolchain}"
+        assert False
+
+    @property
+    def required_tools(self):
+        if self.toolchain == "Apicula":
+            return self._apicula_required_tools
+        elif self.toolchain == "Gowin":
+            return self._gowin_required_tools
+        assert False
+
+    @property
+    def file_templates(self):
+        if self.toolchain == "Apicula":
+            return self._apicula_file_templates
+        elif self.toolchain == "Gowin":
+            return self._gowin_file_templates
+        assert False
+
+    @property
+    def command_templates(self):
+        if self.toolchain == "Apicula":
+            return self._apicula_command_templates
+        elif self.toolchain == "Gowin":
+            return self._gowin_command_templates
+        assert False
+
+    def add_clock_constraint(self, clock, frequency):
+        super().add_clock_constraint(clock, frequency)
+        clock.attrs["keep"] = "true"
+
+    @property
+    def default_clk_constraint(self):
+        if self.default_clk == "sys_clk0":
+            return Clock(self.osc_freq / self.osc_div)
+        # Use the defined Clock resource.
+        return super().default_clk_constraint
+
+    def create_missing_domain(self, name):
+        if name == "sync" and self.default_clk is not None:
+            m = Module()
+            if self.default_clk == "sys_clk0":
+                props = self._osc_props
+                osc = self._oscillator
+                if (not isinstance(self.osc_div, int)
+                    or self.osc_div < props["div_min"]
+                    or self.osc_div > props["div_max"]
+                    or (self.osc_div % props["div_step"]) != 0):
+                    raise ValueError("OSC divider (osc_div) must be an integer between {} and {} in steps of {}, not {!r}".format(
+                        props["div_min"], props["div_max"], props["div_step"], self.osc_div))
+
+                clk_i = Signal()
+                if osc is "OSCZ":
```

Comment:
It's not correct to compare Python strings using `is`.
```suggestion
                if osc == "OSCZ":
```

Reply:
Fixed